### PR TITLE
[Ready] Add whisper TensorRT-LLM

### PR DESCRIPTION
**Pull request by yuekaizhang**
This PR suports whisper [TensorRT-LLM backend](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples/whisper). 

Env: nvcr.io/nvidia/tritonserver:24.09-trtllm-python-py3
Greedy Search Results on an A100 PCIE 80 GB:

Filtering models by id: large-v3
********************************************************************************
Results per dataset:
********************************************************************************
whisper_large/v3 | hf-audio-esb-datasets-test-only-sorted_ami_test: WER = 15.19 %, RTFx = 105.41
whisper_large/v3 | hf-audio-esb-datasets-test-only-sorted_earnings22_test: WER = 10.88 %, RTFx = 244.48
whisper_large/v3 | hf-audio-esb-datasets-test-only-sorted_gigaspeech_test: WER = 10.01 %, RTFx = 210.10
whisper_large/v3 | hf-audio-esb-datasets-test-only-sorted_librispeech_test.clea: WER = 1.96 %, RTFx = 245.05
whisper_large/v3 | hf-audio-esb-datasets-test-only-sorted_librispeech_test.other: WER = 3.80 %, RTFx = 222.43
whisper_large/v3 | hf-audio-esb-datasets-test-only-sorted_spgispeech_test: WER = 2.70 %, RTFx = 289.69
whisper_large/v3 | hf-audio-esb-datasets-test-only-sorted_tedlium_test: WER = 3.94 %, RTFx = 255.01
whisper_large/v3 | hf-audio-esb-datasets-test-only-sorted_voxpopuli_test: WER = 7.60 %, RTFx = 305.96

********************************************************************************
Composite Results:
********************************************************************************
whisper_large/v3: WER = 7.01 %
whisper_large/v3: RTFx = 243.22
********************************************************************************

### CODE REVIEW

**Comment by yuekaizhang** [comment]
@Vaibhavs10  Would you mind helping checking this PR, thanks.

**Comment by nithinraok** [comment]
Is VAD required here? What is the reason to use VAD for whisper?

**Comment by yuekaizhang** [comment]
> Is VAD required here? What is the reason to use VAD for whisper?

Several datasets e.g. tedlium have audios more than 30 seconds.

**Review by Vaibhavs10 — COMMENTED** [review]
I'm quite okay with the PR generally, Just not sure why we use VAD as we do not use this for other backends. I'd generally lean towards keeping the eval logic 1:1 for all backends.

cc: @nithinraok @titu1994

**Comment by yuekaizhang** [comment]
> I'm quite okay with the PR generally, Just not sure why we use VAD as we do not use this for other backends. I'd generally lean towards keeping the eval logic 1:1 for all backends.
> 
> cc: @nithinraok @titu1994

I just follow the faster_whisper [PR](https://github.com/huggingface/open_asr_leaderboard/commit/e4aeb4fe953d44ec57e4dad9be3ce789f3d9db88https://github.com/huggingface/open_asr_leaderboard/blob/main/ctranslate2/run_eval.py#L26) to use VAD. I think faster_whisper used the same vad internally. See https://github.com/SYSTRAN/faster-whisper/blob/master/faster_whisper/transcribe.py#L239.

**Comment by titu1994** [comment]
I don't think vad should be used. We took a lot of effort to make the benchmark consistent and use chunking everywhere, but this gives an advance to whisper

**Comment by yuekaizhang** [comment]
Okay, let me support the chunked long-form function to make consistent.

**Comment by yuekaizhang** [comment]
@Vaibhavs10  Sorry for the delay. Updated results for chunked long-form implementation here.

Filtering models by id: large-v3
********************************************************************************
Results per dataset:
********************************************************************************
whisper_large/v3 | hf-audio-esb-datasets-test-only-sorted_ami_test: WER = 15.27 %, RTFx = 105.48
whisper_large/v3 | hf-audio-esb-datasets-test-only-sorted_earnings22_test: WER = 10.82 %, RTFx = 235.52
whisper_large/v3 | hf-audio-esb-datasets-test-only-sorted_gigaspeech_test: WER = 10.04 %, RTFx = 200.86
whisper_large/v3 | hf-audio-esb-datasets-test-only-sorted_librispeech_test.clea: WER = 1.99 %, RTFx = 230.47
whisper_large/v3 | hf-audio-esb-datasets-test-only-sorted_librispeech_test.other: WER = 3.87 %, RTFx = 210.87
whisper_large/v3 | hf-audio-esb-datasets-test-only-sorted_spgispeech_test: WER = 2.69 %, RTFx = 272.46
whisper_large/v3 | hf-audio-esb-datasets-test-only-sorted_tedlium_test: WER = 4.02 %, RTFx = 240.09
whisper_large/v3 | hf-audio-esb-datasets-test-only-sorted_voxpopuli_test: WER = 7.85 %, RTFx = 286.61

********************************************************************************
Composite Results:
********************************************************************************
whisper_large/v3: WER = 7.07 %
whisper_large/v3: RTFx = 231.53
********************************************************************************

File: tensorrtllm/convert_checkpoint.py [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+
+import tensorrt_llm
+from tensorrt_llm.functional import LayerNormPositionType, LayerNormType
+from tensorrt_llm.models.convert_utils import weight_only_quantize_dict
+from tensorrt_llm.quantization import QuantAlgo
+
+
+def parse_arguments():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--model_dir', type=str, default="assets")
```
Comment (Vaibhavs10) [inline tensorrtllm/convert_checkpoint.py:32]: Can these model checkpoints be on the Hub and directly downloaded from there?

Comment (yuekaizhang) [reply tensorrtllm/convert_checkpoint.py:32]: The convert_checkpoint.py accepts large-v3.pt and generates trt-llm checkpoints. Which one you would like put on the hub?

Comment (yuekaizhang) [reply tensorrtllm/convert_checkpoint.py:32]: https://huggingface.co/yuekai/whisper_turbo_and_distill_tllm_checkpoint/tree/main, like this one, if we store the converted checkpoints, we could remove the convert_checkpoint.py file.

Comment (Vaibhavs10) [reply tensorrtllm/convert_checkpoint.py:32]: Yes exactly like the last one is okay and then we can remove the `convert_checkpoint.py` too. I'd recommend uploading each model in a seperate model repo. We can also move them over to `nvidia` org.

Comment (yuekaizhang) [reply tensorrtllm/convert_checkpoint.py]: Done.

File: requirements/requirements_trtllm.txt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+tiktoken
+jiwer
```
Comment (nithinraok) [inline requirements/requirements_trtllm.txt:2]: why is this required?

File: requirements/requirements_trtllm.txt [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
+tensorrt-llm==0.15.0.dev2024101500
+soundfile
```
Comment (nithinraok) [inline requirements/requirements_trtllm.txt:4]: looks like you are not using soundfile as well

File: tensorrtllm/run_eval.py [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+        start += chunk_size - overlap_size
+    
+    return chunks
+
+def main(args):
+    asr_model = WhisperTRTLLM(engine_dir=args.model_id)
+
+    def benchmark(batch, min_new_tokens=None):
+        # Load audio inputs
+        max_duration, sample_rate, max_batch_size = 30, 16000, 64
```
Comment (nithinraok) [inline tensorrtllm/run_eval.py:142]: use batch_size from args

**Review by nithinraok — APPROVED** [review]
Minor comments. LGTM

File: requirements/requirements_trtllm.txt (hunk 3 above)
Comment (yuekaizhang) [reply requirements/requirements_trtllm.txt:4]: image
It's required by datasets.

File: requirements/requirements_trtllm.txt (hunk 2 above)
Comment (yuekaizhang) [reply requirements/requirements_trtllm.txt:2]: tiktoken for whisper official tokenizer. jiwer for the repo's wer metric computation.

**Comment by yuekaizhang** [comment]
Filtering models by id: large-v3-turbo
********************************************************************************
Results per dataset:
********************************************************************************
whisper_large/v3-turbo | hf-audio-esb-datasets-test-only-sorted_ami_test: WER = 15.73 %, RTFx = 158.87
whisper_large/v3-turbo | hf-audio-esb-datasets-test-only-sorted_earnings22_test: WER = 11.03 %, RTFx = 410.65
whisper_large/v3-turbo | hf-audio-esb-datasets-test-only-sorted_gigaspeech_test: WER = 10.24 %, RTFx = 364.08
whisper_large/v3-turbo | hf-audio-esb-datasets-test-only-sorted_librispeech_test.clea: WER = 1.99 %, RTFx = 422.10
whisper_large/v3-turbo | hf-audio-esb-datasets-test-only-sorted_librispeech_test.other: WER = 3.91 %, RTFx = 377.28
whisper_large/v3-turbo | hf-audio-esb-datasets-test-only-sorted_spgispeech_test: WER = 2.74 %, RTFx = 514.76
whisper_large/v3-turbo | hf-audio-esb-datasets-test-only-sorted_tedlium_test: WER = 6.16 %, RTFx = 455.41
whisper_large/v3-turbo | hf-audio-esb-datasets-test-only-sorted_voxpopuli_test: WER = 8.91 %, RTFx = 543.58

********************************************************************************
Composite Results:
********************************************************************************
whisper_large/v3-turbo: WER = 7.59 %
whisper_large/v3-turbo: RTFx = 418.90
********************************************************************************

Update large-v3-turbo results.

**Review by Vaibhavs10 — APPROVED** [review]
Good to merge for me. I suppose we should run all the TRT LLM compatible Whisper checkpoints right?
